src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -465,6 +426,59 @@ private bool TryRegister(SafeSocketHandle socket, IntPtr handle, out Interop.Err
             return error == Interop.Error.SUCCESS;
         }
 
+        // The JIT is allowed to arbitrarily extend the lifetime of locals, which may retain SocketAsyncContext references,
+        // indirectly preventing Socket instances to be finalized, despite being no longer referenced by user code.
+        // To avoid this, the event handling logic is delegated to a non-inlined processing method.
+        // See discussion: https://github.com/dotnet/runtime/issues/37064
+        // SocketEventHandler holds an on-stack cache of SocketAsyncEngine members needed by the handler method.
+        private struct SocketEventHandler
+        {
+            public bool Shutdown { get; private set; }
+            public Interop.Sys.SocketEvent* Buffer { get; }
+
+            private readonly ConcurrentDictionary<IntPtr, SocketAsyncContextWrapper> _handleToContextMap;
+            private readonly ConcurrentQueue<SocketIOEvent> _eventQueue;
+
+            public SocketEventHandler(SocketAsyncEngine engine)
+            {
+                Shutdown = false;
+                Buffer = engine._buffer;
+                _handleToContextMap = engine._handleToContextMap;
+                _eventQueue = engine._eventQueue;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
The use of aggressiveinlining needs explanation

Reply:
I pushed it now just to check CI results, will update the comments if we stick with this solution. 

Do you think the behavior @tmds quoted  is safe to rely on in long term?
https://github.com/dotnet/corefx/pull/38499#discussion_r293589835
> You can even mark this new method with AggressiveInlinining; the jit does a little bit of explicit nulling on its own for inlinee locals so that inlining doesn't stretch lifetimes. That would catch this case (see dotnet/coreclr#9479).

Reply:
Unless we have a really good reason to use it, we should remove it.

Reply:
Inlining into a `try` block isn't great as the error handling has extra stack/register overheads, and unless it regularly only loops once then any gain from inlining that will be drowned out by the time in the loop?

Reply:
Since its a large loop that's being inlined